packages/data/src/types.ts

```diff
@@ -0,0 +1,32 @@
+export type WPDataFunctionOrGeneratorArray = Array< Function | Generator >;
+export type WPDataFunctionArray = Array< Function >;
```

Comment:
We'll see the linter will report errors here as part of the typescript-eslint recommended rules. [`Function` type is discouraged](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md#default-options):

> Avoid the `Function` type, as it provides little safety for the following reasons:
> - It provides no type safety when calling the value, which means it's easy to provide the wrong arguments.
> - It accepts class declarations, which will fail when called, as they are called without the new keyword.

Example:

```ts
function callAndToLower(f: Function) {
  f().toLowerCase(); // Unsound, may error: `f()` is of type `any`!
}
function safeCallAndToLower(f: () => string) {
  f().toLowerCase(); // All good!
}
// No type errors!
callAndToLower(class { }); // TypeError: Class constructors cannot be invoked without 'new' 
callAndToLower(() => null); // Runtime TypeError: Cannot read property 'toLowerCase' of null

// Sound type errors 👍
safeCallAndToLower(class { });
safeCallAndToLower(() => null);
safeCallAndToLower(() => "Hello!"); // "hello!"
```

FYI @youknowriad